rpc/src/client/subscription.rs

```diff
@@ -358,91 +270,221 @@ impl SubscriptionRouter {
         };
         subs_for_query.remove(id);
     }
+}
 
-    /// Keeps track of a pending unsubscribe request, which can either be
-    /// confirmed or cancelled.
-    pub fn pending_remove(
-        &mut self,
-        req_id: &str,
-        subs_id: &SubscriptionId,
-        query: String,
-        result_tx: ChannelTx<Result<()>>,
-    ) {
-        self.pending_unsubscribe.insert(
-            req_id.to_string(),
-            PendingUnsubscribe {
-                id: subs_id.clone(),
-                query,
-                result_tx,
-            },
-        );
+impl Default for SubscriptionRouter {
+    fn default() -> Self {
+        Self {
+            subscriptions: HashMap::new(),
+        }
     }
+}
+
+#[cfg(feature = "websocket-client")]
+mod two_phase_router {
+    use super::*;
 
-    /// Confirm the pending unsubscribe request for the subscription with the
-    /// given ID.
-    pub async fn confirm_remove(&mut self, req_id: &str) -> Result<()> {
-        match self.pending_unsubscribe.remove(req_id) {
-            Some(mut pending_unsubscribe) => {
-                self.remove(&pending_unsubscribe.id, pending_unsubscribe.query.clone());
-                Ok(pending_unsubscribe.result_tx.send(Ok(())).await?)
+    /// A subscription router that can manage pending subscribe and unsubscribe
+    /// requests, as well as their confirmation/cancellation.
+    ///
+    /// This is useful in instances where the underlying transport is complex,
+    /// e.g. WebSocket connections, where many messages are multiplexed on the
+    /// same communication line. In such cases, a response from the remote
+    /// endpoint immediately after a subscribe/unsubscribe request may not be
+    /// relevant to that request.
+    #[derive(Debug)]
+    pub struct TwoPhaseSubscriptionRouter {
```

Comment:
Don't love the name. Not sure what a better one would be but probably something more descriptive of the functionality. I guess it's kind of like a cache? I also still cant help but wonder if we can get rid of it per https://github.com/informalsystems/tendermint-rs/pull/576#discussion_r493159561.